### PR TITLE
iox-#1542 Support Bazel build system

### LINKED ISSUE #1542: Support Bazel build system (as alternative to CMake)
## Brief feature description

Add support for building iceoryx with [Bazel](https://bazel.build/) build tool (as alternative build system to  [CMake](https://cmake.org/)).

## Detailed information

[Bazel](https://bazel.build/) provides advanced build features, supporting hermetic, reproducible and scalable builds. 

[Bazel](https://bazel.build/) could be added additionally - without breaking existing [CMake](https://cmake.org/) builds. 

Additional build files (`WORKSPACE.bazel` and `BUILD.bazel`) are needed as well as (maybe) some bazel rules (`.bzl`) located in a `bazel` subdirectory (like already existing `cmake` subdirectory).

### PR DESCRIPTION
**Pull request by kwallner**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

### Building

Building as described in `doc/website/getting-started/installation.md` with:
```
bazel build //...
```

Build will use `gcc`. Other compilers can be selected using environment variables:
```
CC=clang bazel build //...
```

### Testing
Running all tests with:
```
bazel test //...
```

#### 1. Bazel version is not defined

Would be an option to set a specific bazel version in `.bazelversion` file.

#### 2. C++-Standard is set to C++14

Used `.bazelrc` to set compiler flags for c++14.
Changing will require to edit `.bazelrc`.

#### 3. Only linux (ubuntu 20.04) tested

Extending to different operating systems would be an follow up issue.
Maybe should be created right away.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1542

### CODE REVIEW

**Review by elfenpiff — CHANGES_REQUESTED** [review]
I support this new feature completely but two questions/things are still open for me:

1. Who is responsible for the bazel setup and takes care of additional bazel issues? @kwallner would you be willing to respond quickly to upcoming bazel issues?

2. We have at the moment no verification by the CI that this bazel setup is actually working. Could you add a bazel CI target for ubuntu-latest where everything is build.

3. For which platforms we want to provide bazel support?

**Comment by dkroenke** [comment]
@elfenpiff 

> 1. Who is responsible for the bazel setup and takes care of additional bazel issues? @kwallner would you be willing to respond quickly to upcoming bazel issues?

That will be me.

> 2. We have at the moment no verification by the CI that this bazel setup is actually working. Could you add a bazel CI target for ubuntu-latest where everything is build.

The CI target will be added with this MR, Karl just made the intial commits ready.
We will add a sanity check job on Ubuntu that builds and test with bazel.

> 3. For which platforms do we want to provide bazel support?

Ubuntu only (20.04).

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 #include "iceoryx_hoofs/cxx/type_traits.hpp"
```
Comment (dkroenke) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:20]: Fix formatting

```suggestion
#include "iceoryx_hoofs/cxx/optional.hpp"
#include "iceoryx_hoofs/cxx/type_traits.hpp"
```

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 2]
```diff
@@ -19,7 +19,7 @@
...
 #include "iceoryx_hoofs/internal/cxx/string_internal.hpp"
-#include "optional.hpp"
+#include "iceoryx_hoofs/cxx/optional.hpp"
```
Comment (dkroenke) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:22]:
```suggestion
```

**Review by dkroenke — CHANGES_REQUESTED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 2 above)
Comment (kwallner) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp]: @dkroenke Did already changed this. Please check if the result matches your expectations.

**Comment by dkroenke** [comment]
@kwallner 
Regarding the open points to discuss:

> 1. Bazel version is not defined
Would be an option to set a specific bazel version in .bazelversion file.

Wouldn't this not make our project harder to integrate into other bazel workspaces?
For me it make sense to support the latest (stable) version of bazel that we should test in the CI.
According to the bazel [Bazel versioning](https://docs.bazel.build/versions/main/versioning.html) this would be currently 5.2 to use in the CI.
Support for older versions is not planned.
For now we can try to live without a bazel version file for more flexibility, if we face increased effort by maintaining multiple versions we can still fix the version.

> 2. C++-Standard is set to C++14
Used .bazelrc to set compiler flags for c++14.
Changing will require to edit .bazelrc.

For now it is ok to enforce here C++14 as we do the same in CMake. We do not change the C++ standard that often so that maintaining this file should not be a problem.

> 3. Only linux (ubuntu 20.04) tested
Extending to different operating systems would be an follow up issue.
Maybe should be created right away.

We have to consider that we already face considerable build times on the GitHub Actions with the existing jobs due to increased usage of Runners for etc.
MacOS and Windows support is mainly driven by the ROS 2 integration where Bazel is not used as build system.
The MacOS runners are already in excessive use so that i would here more rely on the community if the build on MacOS faces a problem.


Possible follow-ups on the following questions:

- [ ] Code-coverage
- [ ] Adress-sanitizer and friends in bazel

**Comment by dkroenke** [comment]
@kwallner You need to add the `bazel-*` directories to .gitignore file.

**Comment by kwallner** [comment]
@elfenpiff 

"Code Linting" is caused by failed request to https://github.com/eclipse-iceoryx/iceoryx/blob/master/BUILD .
This file will exist after the merge. 

How to handle this?

File: doc/website/advanced/installation-guide-for-contributors.md [hunk 3]
```diff
@@ -142,10 +142,41 @@ machine. With `./tools/scripts/ice_env.sh` one can create a
...
+
+For formatting and linting this command will do the job:
+
+```bash
+cd iceoryx
+bazel run //:buildifier_lint
+```
+
+The CI will check for the correct formatting and linting.
+See the [BUILD](https://github.com/eclipse-iceoryx/iceoryx/blob/master/BUILD) file
```
Comment (dkroenke) [inline doc/website/advanced/installation-guide-for-contributors.md:181]:
```suggestion
See the `BUILD.bazel` file
```

**Comment by kwallner** [comment]
@dkroenke 

Don't understand checklist task "All touched (C/C++) source code files are added to ./clang-tidy-diff-scans.txt". What is the actual TODO here?

**Comment by dkroenke** [comment]
> @dkroenke
> 
> Don't understand checklist task "All touched (C/C++) source code files are added to ./clang-tidy-diff-scans.txt". What is the actual TODO here?

@kwallner this check affects only the C++ code in iceoryx, since we are only touching the include paths for the internal API (and correct it) this is out of scope here.

**Comment by dkroenke** [comment]
LGTM from my side! Many thanks @kwallner for this great contribution. :clap:

**Review by dkroenke — APPROVED** [review]

**Comment by dkroenke** [comment]
Folow-up topics for Bazel are tracked here: https://github.com/eclipse-iceoryx/iceoryx/issues/1547

**Review by elfenpiff — APPROVED** [review]

**Pull request merged**